File: LeetCode/iter1/c++/728.self-dividing-numbers.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	vector<int> selfDividingNumbers(int left, int right) {
		vector<int> v;
		for (int i = left; i <= right; i++) {
			if (whetherDividing(i)) {
				v.push_back(i);
			}
		}
		return v;
	}
private:
	bool whetherDividing(int num) {
		bool ret = true;
		int tmpNum = num, digit;
		while (tmpNum != 0) {
			digit = tmpNum % 10;
			tmpNum /= 10;
			if (digit == 0 || num % digit != 0) {
				ret = false;
				break;
			}
		}

		return ret;
	}
};

void testPrint(vector<int> v) {
	for (int num : v) {
		cout << num << " ";
	}
	cout << endl;
}

int main(void)
{
	Solution s;
	testPrint(s.selfDividingNumbers(1, 22));

	return 0;
}

File: LeetCode/iter1/java/728.self-dividing-numbers.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Solution {

    public static void main(String[] args) {
        System.out.println(selfDividingNumbers(1, 22));
    }

    public static List<Integer> selfDividingNumbers(int left, int right) {
        List<Integer> ret = new ArrayList<>();
        for (int i = left; i <= right; i++) {
            if (whetherSelfDividing(i)) {
                ret.add(i);
            }
        }

        return ret;
    }

    private static boolean whetherSelfDividing(int num) {
        if (num == 0) {
            return false;
        }
        boolean ret = true;
        int tmpNum = num, digit;
        while (tmpNum != 0) {
            digit = tmpNum % 10;
            tmpNum /= 10;
            if (digit == 0 || num % digit != 0) {
                ret = false;
                break;
            }
        }

        return ret;
    }
}